### PR TITLE
[proto] Added functional affine_segmentation_mask op

### PR DESCRIPTION
**Pull request by vfdev-5**
Related to #5514

## Description:
- Added functional `affine_segmentation_mask` op
- Added tests

### Results on synthetic images/bboxes/segm mask:

<details>
<summary>
Code
</summary>

```python
import numpy as np

import torch
import torchvision
from torchvision.prototype import features
from torchvision.prototype.transforms.functional import affine_image_tensor, affine_bounding_box, affine_segmentation_mask

size = (64, 76)
# xyxy format
in_boxes = [
    [10, 15, 25, 35],
    [50, 5, 70, 22],
    [45, 46, 56, 62],
]
labels = [1, 2, 3]

im1 = 255 * np.ones(size + (3, ), dtype=np.uint8)
mask = np.zeros(size, dtype=np.int64)
for in_box, label in zip(in_boxes, labels):
    im1[in_box[1]:in_box[3], in_box[0]:in_box[2], :] = (127, 127, 127)
    mask[in_box[1]:in_box[3], in_box[0]:in_box[2]] = label
    
t_im1 = torch.tensor(im1).permute(2, 0, 1).view(1, 3, *size)

in_boxes = features.BoundingBox(
    in_boxes, format=features.BoundingBoxFormat.XYXY, image_size=size
)
in_mask = features.SegmentationMask(torch.tensor(mask)).view(1, *size)
    
angle = 34
scale = 0.9
t = (-6, 7)
shear = (1, 2)

out_boxes = affine_bounding_box(
    in_boxes, 
    in_boxes.format,
    in_boxes.image_size,
    angle,
    t,
    scale,
    shear,
)
print(out_boxes)

out_mask = affine_segmentation_mask(
    in_mask, 
    angle,
    t,
    scale,
    shear    
)

t_im2 = affine_image_tensor(t_im1, angle, t, scale, shear)


import cv2
import matplotlib.pyplot as plt
%matplotlib inline


plt.figure(figsize=(14, 10))

plt.subplot(2,3,1)
plt.title("Input image + bboxes")
r1 = t_im1[0, ...].permute(1, 2, 0).contiguous().cpu().numpy()
for in_box in in_boxes:    
    r1 = cv2.rectangle(r1, (in_box[0].item(), in_box[1].item()), (in_box[2].item(), in_box[3].item()), (255, 127, 0))
plt.imshow(r1)


plt.subplot(2,3,2)
plt.title("Input segm mask")
plt.imshow(in_mask[0, :, :].cpu().numpy())


plt.subplot(2,3,3)
plt.title("Input image + bboxes + segm mask")
plt.imshow(r1, alpha=0.5)
plt.imshow(in_mask[0, :, :].cpu().numpy(), alpha=0.75)


plt.subplot(2,3,4)
plt.title("Output image + bboxes")
r2 = t_im2[0, ...].permute(1, 2, 0).contiguous().cpu().numpy()
for out_box in out_boxes:
    out_box = np.round(out_box.cpu().numpy()).astype("int32")
    r2 = cv2.rectangle(r2, (out_box[0], out_box[1]), (out_box[2], out_box[3]), (255, 127, 0), 0)
plt.imshow(r2)


plt.subplot(2,3,5)
plt.title("Output segm mask")
plt.imshow(out_mask[0, :, :].cpu().numpy())

plt.subplot(2,3,6)
plt.title("Output image + bboxes + segm mask")
plt.imshow(r2, alpha=0.5)
plt.imshow(out_mask[0, :, :].cpu().numpy(), alpha=0.75)
```
</details>

image

### Compare to albumentations:

TL;DR: results do not match due to missing offset and https://github.com/opencv/opencv/issues/11784

image

<details>
<summary>
Code
</summary>

```python
# pip install albumentations
# pip install git+https://github.com/pytorch/data

import numpy as np
import cv2
import albumentations
from albumentations.augmentations.geometric.functional import shift_scale_rotate

import torch
import torchvision
from torchvision.prototype import features
from torchvision.prototype.transforms.functional import affine_segmentation_mask

print(torch.__version__)
print(torchvision.__version__)
print(albumentations.__version__)


size = (64, 64)
# xyxy format
in_boxes = [
    [50, 5, 70, 22],
    [size[1] // 2 - 10, size[0] // 2 - 10, size[1] // 2 + 10, size[0] // 2 + 10],
    [1, 1, 5, 5],
]
labels = [1, 2, 3]

im1 = 255 * np.ones(size + (3, ), dtype=np.uint8)
mask = np.zeros(size, dtype=np.int64)
for in_box, label in zip(in_boxes, labels):
    im1[in_box[1]:in_box[3], in_box[0]:in_box[2], :] = (127, 127, 127)
    mask[in_box[1]:in_box[3], in_box[0]:in_box[2]] = label


# Params
angle = 63
scale = 0.89
dx = 0.12
dy = 0.23


# https://github.com/albumentations-team/albumentations/blob/89a675cbfb2b76f6be90e7049cd5211cb08169a5/albumentations/augmentations/geometric/transforms.py#L81
albu_out_mask = shift_scale_rotate(mask, -angle, scale, dx, dy, cv2.INTER_NEAREST, cv2.BORDER_CONSTANT, 0)

# Using offset for images
# https://github.com/opencv/opencv/issues/11784
offset = 0.5
center = (size[1] / 2 - offset, size[0] / 2 - offset)
m = cv2.getRotationMatrix2D(center, -angle, scale=scale)
m[0, 2] += dx * size[1]
m[1, 2] += dy * size[0]
cv2_out_mask = cv2.warpAffine(mask, m, dsize=size[::-1], flags=cv2.INTER_NEAREST, borderValue=0, borderMode=0)

in_mask = features.SegmentationMask(torch.tensor(mask)).view(1, *size)

out_mask = affine_segmentation_mask(
    in_mask,
    angle,
    (dx * size[1], dy * size[0]),
    scale,
    shear=(0, 0)
)


import matplotlib.pyplot as plt
%matplotlib inline

plt.figure(figsize=(20, 15))
plt.subplot(141)
plt.title("Input mask")
plt.imshow(mask)
plt.subplot(142)
plt.title("Output mask by torchvision")
plt.imshow(np_out_mask)
plt.subplot(143)
plt.title("Output mask by albumentations")
plt.imshow(albu_out_mask)
plt.subplot(144)
plt.title("Output mask diff: torchvision - albumentations")
plt.imshow(np_out_mask - albu_out_mask)


plt.figure(figsize=(20, 15))
plt.subplot(141)
plt.title("Input mask")
plt.imshow(mask)
plt.subplot(142)
plt.title("Output mask by torchvision")
plt.imshow(np_out_mask)
plt.subplot(143)
plt.title("Output mask by offsetted cv2 affine warp")
plt.imshow(cv2_out_mask)
plt.subplot(144)
plt.title("Output mask diff: torchvision - offsetted cv2 affine warp")
plt.imshow(np_out_mask - cv2_out_mask)
```

image

</details>

### CODE REVIEW

File: torchvision/prototype/transforms/functional/_geometry.py [hunk 1]
```diff
@@ -229,6 +229,27 @@ def affine_bounding_box(
...
+
+    center_f = [0.0, 0.0]
+    if center is not None:
+        _, height, width = get_dimensions_image_tensor(img)
+        # Center values should be in pixel coordinates but translated such that (0, 0) corresponds to image center.
+        center_f = [1.0 * (c - s * 0.5) for c, s in zip(center, [width, height])]
+
+    translate_f = [1.0 * t for t in translate]
+    matrix = _get_inverse_affine_matrix(center_f, angle, translate_f, scale, shear)
+    return _FT.affine(img, matrix, interpolation=InterpolationMode.NEAREST.value, fill=None)
```
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_geometry.py:250]: Maybe I'm missing something here, but this seems copy-pasted from `affine_image_tensor`. Why not call that directly?

File: test/test_prototype_transforms_functional.py [hunk 2]
```diff
@@ -427,3 +461,74 @@ def test_correctness_affine_bounding_box_on_fixed_input(device):
...
+
+@pytest.mark.parametrize("device", cpu_and_gpu())
+def test_correctness_affine_segmentation_mask_on_fixed_input(device):
+    # Check transformation against known expected output
+    # Rotate 90 degrees and scale
+    mask = torch.zeros(1, 32, 32, dtype=torch.long, device=device)
+    mask[0, 2:10, 2:10] = 1
+    mask[0, 32 - 9 : 32 - 3, 3:9] = 2
+    mask[0, 1:11, 32 - 11 : 32 - 1] = 3
+    mask[0, 16 - 4 : 16 + 4, 16 - 4 : 16 + 4] = 4
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply test/test_prototype_transforms_functional.py:526]: This is a data creation. I'll leave a comment to make it more clear.

File: test/test_prototype_transforms_functional.py [hunk 3]
```diff
@@ -427,3 +461,74 @@ def test_correctness_affine_bounding_box_on_fixed_input(device):
...
+
+    expected_mask = torch.rot90(mask, k=-1, dims=(-2, -1))
+    expected_mask = torch.nn.functional.interpolate(expected_mask[None, ...].float(), size=(64, 64), mode="nearest")
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply test/test_prototype_transforms_functional.py:529]: This is just a shorter notation for unsqueeze...
It is rather common syntax see for example: https://github.com/pytorch/vision/blob/053e7ebd1b50b7a506cfd6e2389043ec05414b1c/torchvision/transforms/functional_tensor.py#L345-L350

File: test/test_prototype_transforms_functional.py [hunk 4]
```diff
@@ -138,6 +138,22 @@ def make_one_hot_labels(
...
+    size = size or torch.randint(16, 33, (2,)).tolist()
+    shape = (*extra_dims, 1, *size)
+    data = make_tensor(shape, low=0, high=max_value, dtype=dtype)
+    return features.SegmentationMask(data)
+
+
+def make_segmentation_masks(
+    image_sizes=((32, 32), (32, 42)),
+    dtypes=(torch.long,),
+    extra_dims=((), (4,)),
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply test/test_prototype_transforms_functional.py:151]: Affine op does not work on 5D tensors, only 3D and 4D:
```
import torch
from torchvision.prototype.transforms.functional import affine_image_tensor

t_im2 = affine_image_tensor(torch.rand(2, 3, 3, 32, 32), 45, (0, 0), 1.0, (0.0, 0.0))
>
RuntimeError: grid_sampler(): expected 4D or 5D input and grid with same number of dimensions, but got input with sizes [2, 3, 3, 32, 32] and grid with sizes [2, 32, 32, 2]
```

I think it was not required when porting the op from pil to tensor. However, it is not complicated to make it work...

File: test/test_prototype_transforms_functional.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply test/test_prototype_transforms_functional.py:526]: The test checks F.affine_segmentation_mask with a degenerate configuration vs how it could be done using pure torch ops on cpu and cuda.
Input data does not play much role just it shouldn't be trivial as a single mask somewhere such that it would disappear after the transformation.

File: test/test_prototype_transforms_functional.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_transforms_functional.py:151]: I think the plan was to support arbitrary batch dimensions. cc @datumbox to make sure. The implementation should be similar to what we do for bounding boxes: squash all batch dimensions, perform the operation, and reshape to the original shape.

File: test/test_prototype_transforms_functional.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_transforms_functional.py:530]: Hm, I haven't come across it before. Is this documented somewhere? I couldn't find anything in the PyTorch documentation.

File: test/test_prototype_transforms_functional.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply test/test_prototype_transforms_functional.py:151]: Let's enable ND support for images/masks in a follow-up PR

File: test/test_prototype_transforms_functional.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_transforms_functional.py:530]: PyTorch aims for Array API compatibility rather than `numpy` compatibility. As of now, this is part of the specification although there is a request for it data-apis/array-api#360. 

Given that this seems totally undocumented on the PyTorch side, I would prefer not to rely on this syntax given that we are talking about 2 characters here and the alternative is documented.

File: test/test_prototype_transforms_functional.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_transforms_functional.py:151]: I'm ok with that given that we also need to touch `affine_image_tensor`.

**Review by datumbox — APPROVED** [review]
@vfdev-5 LGTM on the kernel side.

@pmeier Are your concerns on the test side covered or you would suggest more changes?

File: test/test_prototype_transforms_functional.py [hunk 5]
```diff
@@ -138,6 +138,22 @@ def make_one_hot_labels(
         yield make_one_hot_label(extra_dims_)
 
 
+def make_segmentation_mask(size=None, *, max_value=80, extra_dims=(), dtype=torch.uint8):
```
Comment (pmeier) [inline test/test_prototype_transforms_functional.py:141]: 1. Why `max_value=80`? Shouldn't that be the max value of the `dtype`?
2. In `make_segmentation_masks` the default dtype is `torch.long`. Should we align the two?

File: test/test_prototype_transforms_functional.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_transforms_functional.py:151]: Could you add it the other `extra_dims` here and explicitly use the reduced set in the affine tests?

File: test/test_prototype_transforms_functional.py [hunk 6]
```diff
@@ -138,6 +138,22 @@ def make_one_hot_labels(
...
+    size = size or torch.randint(16, 33, (2,)).tolist()
+    shape = (*extra_dims, 1, *size)
+    data = make_tensor(shape, low=0, high=max_value, dtype=dtype)
+    return features.SegmentationMask(data)
+
+
+def make_segmentation_masks(
+    image_sizes=((32, 32), (32, 42), (38, 24)),
```
Comment (pmeier) [inline test/test_prototype_transforms_functional.py:149]: Any reason to not use the same values as in `make_images`?

File: test/test_prototype_transforms_functional.py [hunk 7]
```diff
@@ -248,6 +264,24 @@ def affine_bounding_box():
         )
 
 
+@register_kernel_info_from_sample_inputs_fn
+def affine_segmentation_mask():
+    for image, angle, translate, scale, shear in itertools.product(
+        make_images(extra_dims=()),
```
Comment (pmeier) [inline test/test_prototype_transforms_functional.py:270]: 1. This should be `make_segmentation_masks`, right?
2. We should also test the batch case. If https://github.com/pytorch/vision/pull/5613/files#r828077791 is addressed we can remove the special casing here.

```suggestion
        make_images(extra_dims=((), (4,))),
```

File: test/test_prototype_transforms_functional.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_transforms_functional.py:526]: @vfdev-5 could have another look into this? Unless I'm missing something in my comment above, I think we can simplify this test so it is easier to follow.

**Review by pmeier — COMMENTED** [review]
I've added some more test related comments.

File: test/test_prototype_transforms_functional.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply test/test_prototype_transforms_functional.py:526]: In your suggestion with a single off-center rectangle we test the op on a binary but here I check on multi-class mask. It is a detail, but I prefer to ensure that multi-class mask is transformed correctly.

File: test/test_prototype_transforms_functional.py (hunk 5 above)
Comment (vfdev-5) [reply test/test_prototype_transforms_functional.py:141]: max_value is the number of classes in the mask, for example COCO has 81 classes (with BG)

File: test/test_prototype_transforms_functional.py (hunk 6 above)
Comment (vfdev-5) [reply test/test_prototype_transforms_functional.py:149]: I'd say there is no particular reason, just prefer even size images.
What was the reason for you to use `(16, 16), (7, 33), (31, 9)` ?

File: test/test_prototype_transforms_functional.py (hunk 7 above)
Comment (vfdev-5) [reply test/test_prototype_transforms_functional.py:270]: > This should be make_segmentation_masks, right?

yes, let's `make_segmentation_masks`.

File: test/test_prototype_transforms_functional.py (hunk 5 above)
Comment (pmeier) [reply test/test_prototype_transforms_functional.py:141]: Ah, ok. Maybe we should rename it to

```suggestion
def make_segmentation_mask(size=None, *, num_categories=80, extra_dims=(), dtype=torch.uint8):
```

then?

File: test/test_prototype_transforms_functional.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_transforms_functional.py:526]: Isn't a multi-class mask already tested by the reference test? In my understanding this test is a "smoke test" with a simple, easily verifiable example. What did you have in mind for this test?

[the remark this answers is not in the document]
Comment (vfdev-5) [reply test/test_prototype_transforms_functional.py:526]: The goal for this test to ensure correctness on a "simple" case and check cpu/cuda devices as adding cuda to `test_correctness_affine_segmentation_mask` would make CI time too large.

[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_transforms_functional.py:526]: Got it, I missed that. Could you add this as a comment to the test? Maybe we can still reduce the number of masks to 2?

File: test/test_prototype_transforms_functional.py (hunk 6 above)
Comment (vfdev-5) [reply test/test_prototype_transforms_functional.py:149]: OK, i'll put your numbers as I have no neither strong opinion. Just a thought, I agree that odd numbers looks like corner cases but odd sized images are less frequent to my opinion. It would make more sense then to test all combinations to ensure that the op works on square, rectangular and even/odd sized images...
